File: python/exercises/practice/unit_test_3e691344_7a83_46f2_8add_1df29b15d8a9/network_router/.meta/network_router.py
import heapq
import threading
from collections import defaultdict

class NetworkRouter:
    def __init__(self, N, graph):
        self.N = N
        self.graph = defaultdict(list)
        self.link_info = {}
        self.load_lock = threading.Lock()
        
        # Initialize graph and link information
        for u, v, latency, capacity in graph:
            self.graph[u].append((v, latency, capacity))
            self.graph[v].append((u, latency, capacity))
            self.link_info[(u, v)] = {'latency': latency, 'capacity': capacity, 'load': 0}
            self.link_info[(v, u)] = {'latency': latency, 'capacity': capacity, 'load': 0}

    def _get_effective_latency(self, u, v):
        with self.load_lock:
            info = self.link_info[(u, v)]
            load_factor = (info['load'] / info['capacity']) ** 2
            return info['latency'] * (1 + load_factor)

    def find_path(self, S, D):
        if S == D:
            return [S]
            
        # Dijkstra's algorithm with priority queue
        heap = []
        heapq.heappush(heap, (0, S, [S]))
        visited = set()
        
        while heap:
            total_latency, current, path = heapq.heappop(heap)
            
            if current == D:
                # Update loads for the chosen path
                with self.load_lock:
                    for i in range(len(path)-1):
                        u, v = path[i], path[i+1]
                        self.link_info[(u, v)]['load'] += 1
                        self.link_info[(v, u)]['load'] += 1
                return path
                
            if current in visited:
                continue
                
            visited.add(current)
            
            for neighbor, latency, capacity in self.graph[current]:
                if neighbor not in visited:
                    effective_latency = self._get_effective_latency(current, neighbor)
                    new_total = total_latency + effective_latency
                    new_path = path + [neighbor]
                    heapq.heappush(heap, (new_total, neighbor, new_path))
        
        return []  # No path found

    def update_loads(self, delta):
        with self.load_lock:
            for key in self.link_info:
                new_load = max(0, self.link_info[key]['load'] + delta)
                self.link_info[key]['load'] = new_load

    def get_link_load(self, u, v):
        with self.load_lock:
            return self.link_info[(u, v)]['load']

File: go/exercises/practice/unit_test_f160bd69_ea56_41e5_ab46_fcc203324a6c/network_router/.meta/network_router.go
package networkrouter

import (
	"container/heap"
	"math"
)

type Node struct {
	ID int
}

type Edge struct {
	Source      int
	Destination int
	Latency     int
	Bandwidth   int
}

type ContentRequest struct {
	UserID            int
	ContentID         int
	SourceServer      int
	DestinationServer int
	BandwidthRequired int
}

type NetworkGraph struct {
	Nodes []Node
	Edges []Edge
}

// Helper struct to store path information during search
type pathNode struct {
	nodeID    int
	latency   int
	bandwidth int
	path      []int
}

// Priority queue implementation for Dijkstra's algorithm
type priorityQueue []*pathNode

func (pq priorityQueue) Len() int           { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool { return pq[i].latency < pq[j].latency }
func (pq priorityQueue) Swap(i, j int)      { pq[i], pq[j] = pq[j], pq[i] }
func (pq *priorityQueue) Push(x interface{}) { *pq = append(*pq, x.(*pathNode)) }
func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

// findPath finds the optimal path considering bandwidth constraints
func findPath(graph NetworkGraph, start, end int, requiredBandwidth int, usedBandwidth map[string]int) ([]int, bool) {
	// Initialize distances
	distances := make(map[int]int)
	paths := make(map[int][]int)
	pq := &priorityQueue{}
	heap.Init(pq)

	// Initialize start node
	heap.Push(pq, &pathNode{
		nodeID:    start,
		latency:   0,
		bandwidth: math.MaxInt32,
		path:      []int{start},
	})
	distances[start] = 0

	// Create adjacency list for faster lookup
	adjList := make(map[int][]Edge)
	for _, edge := range graph.Edges {
		adjList[edge.Source] = append(adjList[edge.Source], edge)
	}

	for pq.Len() > 0 {
		current := heap.Pop(pq).(*pathNode)

		if current.nodeID == end {
			return current.path, true
		}

		// Skip if we've found a better path already
		if current.latency > distances[current.nodeID] {
			continue
		}

		// Process neighbors
		for _, edge := range adjList[current.nodeID] {
			remainingBandwidth := edge.Bandwidth - usedBandwidth[getEdgeKey(edge.Source, edge.Destination)]
			if remainingBandwidth < requiredBandwidth {
				continue
			}

			newLatency := current.latency + edge.Latency
			if dist, exists := distances[edge.Destination]; !exists || newLatency < dist {
				distances[edge.Destination] = newLatency
				newPath := make([]int, len(current.path))
				copy(newPath, current.path)
				newPath = append(newPath, edge.Destination)
				paths[edge.Destination] = newPath

				heap.Push(pq, &pathNode{
					nodeID:    edge.Destination,
					latency:   newLatency,
					bandwidth: min(current.bandwidth, remainingBandwidth),
					path:      newPath,
				})
			}
		}
	}

	return nil, false
}

func getEdgeKey(src, dst int) string {
	return string(rune(src)) + "-" + string(rune(dst))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// OptimizeNetwork implements the main network optimization logic
func OptimizeNetwork(networkGraph NetworkGraph, requests []ContentRequest) (map[int][]int, []int) {
	routedRequests := make(map[int][]int)
	rejectedRequests := []int{}
	usedBandwidth := make(map[string]int)

	// Sort requests by bandwidth required (larger requests first)
	sortedRequests := make([]ContentRequest, len(requests))
	copy(sortedRequests, requests)
	for i := 0; i < len(sortedRequests)-1; i++ {
		for j := i + 1; j < len(sortedRequests); j++ {
			if sortedRequests[i].BandwidthRequired < sortedRequests[j].BandwidthRequired {
				sortedRequests[i], sortedRequests[j] = sortedRequests[j], sortedRequests[i]
			}
		}
	}

	// Process each request
	for _, request := range sortedRequests {
		path, found := findPath(networkGraph, request.SourceServer, request.DestinationServer,
			request.BandwidthRequired, usedBandwidth)

		if !found {
			rejectedRequests = append(rejectedRequests, request.UserID)
			continue
		}

		// Update bandwidth usage for the path
		for i := 0; i < len(path)-1; i++ {
			edgeKey := getEdgeKey(path[i], path[i+1])
			usedBandwidth[edgeKey] += request.BandwidthRequired
		}

		routedRequests[request.UserID] = path
	}

	return routedRequests, rejectedRequests
}